src/inference/AdvancedSMC.jl

```diff
@@ -147,72 +224,44 @@ function step(model, spl::Sampler{<:PG}, vi::VarInfo)
     ## pick a particle to be retained.
     Ws, _ = weights(particles)
     indx = randcat(Ws)
-    push!(spl.info[:logevidence], particles.logE)
-
-    return particles[indx].vi, true
-end
-
-function sample(  model::Model,
-                  alg::PG;
-                  save_state=false,         # flag for state saving
-                  resume_from=nothing,      # chain to continue
-                  reuse_spl_n=0             # flag for spl re-using
-                )
-
-    spl = reuse_spl_n > 0 ?
-          resume_from.info[:spl] :
-          Sampler(alg)
-    if resume_from != nothing
-        spl.selector = resume_from.info[:spl].selector
-    end
-    @assert typeof(spl.alg) == typeof(alg) "[Turing] alg type mismatch; please use resume() to re-use spl"
-
-    n = reuse_spl_n > 0 ?
-        reuse_spl_n :
-        alg.n_iters
-    samples = Vector{Sample}()
-
-    ## custom resampling function for pgibbs
-    ## re-inserts reteined particle after each resampling step
-    time_total = zero(Float64)
 
-    vi = resume_from == nothing ?
-        empty!(VarInfo(model)) :
-        resume_from.info[:vi]
+    # Extract the VarInfo from the retained particle.
+    params = tonamedtuple(spl.state.vi)
+    spl.state.vi = particles[indx].vi
+    lp = getlogp(spl.state.vi)
 
-    pm = nothing
-    PROGRESS[] && (spl.info[:progress] = ProgressMeter.Progress(n, 1, "[PG] Sampling...", 0))
-
-    for i = 1:n
-        time_elapsed = @elapsed vi, _ = step(model, spl, vi)
-        push!(samples, Sample(vi))
-        samples[i].value[:elapsed] = time_elapsed
-
-        time_total += time_elapsed
+    # update the master vi.
+    return transition(params, lp, particles.logE, 1.0)
+end
 
-        if PROGRESS[] && spl.selector.tag == :default
-            ProgressMeter.next!(spl.info[:progress])
-        end
-    end
+function sample_end!(
+    ::AbstractRNG,
+    ::Model,
+    spl::Sampler{<:ParticleInference},
+    N::Integer,
+    ts::Vector{ParticleTransition};
+    kwargs...
+)
+    # Set the default for resuming the sampler.
+    resume_from = get(kwargs, :resume_from, nothing)
 
-    @info("[PG] Finished with")
-    @info("  Running time    = $time_total;")
+    # Exponentiate the average log evidence.
+    loge = exp(mean([t.le for t in ts]))
```

Comment:
Consider using
```julia
loge = exp(mean(t.le for t in ts))
```
instead.

Reply:
Maybe also consider renaming the variable. `loge` is actually not what it is. Better `avgevidence` ?

Reply:
Actually, why do we do the `exp` here. I think it is uneccesary.

Reply:
I believe there was such a discussion long time ago but I cannot find which issue was it. @yebai do you happen to remember it?

Reply:
I can't find it either. I remember asking which value we should store and I believe this was the one we agreed to, but I agree the name in the code is not super helpful.

Reply:
I've moved everything over to not doing all this weird `exp` -> `log` -> `exp` stuff we were doing.

Reply:
Do we have any test on `loge`? Just trying to make sure we didn't screw up anything.

Reply:
Resolved by https://github.com/TuringLang/Turing.jl/pull/793#issuecomment-529061951